### ISSUE TITLE
Gtk3 warnings on stderr

### ISSUE THREAD

**Issue by mcepl** [vim/vim#1149]
I have built vim 8.0.022 on RHEL-7 with Gtk3 (gtk3-3.14.13-20.el7.x86_64) and everything works nice (thank you!) except I get plenty of warnings on stderr (even when running `gvim -N -u NONE`):

```
(gvim:10610): Gtk-WARNING **: GtkNotebook 0x28001e0 is mapped
but visible child GtkEventBox 0x2802150 is not mapped
```

According to my colleagues more knowledgeable about Gtk internals (but not willing to touch vim source code :()

> Usually it means that something is manipulating the visibility of children manually, or it's blocking some signal emission.

More discussion on this topic is on https://groups.google.com/d/topic/vim_dev/VfDsEUegJSA/discussion, but it seems to me that the answer “we don’t care” is not good enough. I quite often run gvim from the command line and all those warnings bother me (and alias `gvim 2>/dev/null` seems too low).

The output of `gvim --version` is:

```
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Oct  5 2016 09:05:27)
Included patches: 1-22
Modified by <bugzilla@redhat.com>
Compiled by <bugzilla@redhat.com>
Huge version with GTK3 GUI.  Features included (+) or not (-):
+acl             +file_in_path    +mouse_sgr       +tag_old_static
+arabic          +find_in_path    -mouse_sysmouse  -tag_any_white
+autocmd         +float           +mouse_urxvt     -tcl
+balloon_eval    +folding         +mouse_xterm     +termguicolors
+browse          -footer          +multi_byte      +terminfo
++builtin_terms  +fork()          +multi_lang      +termresponse
+byte_offset     +gettext         -mzscheme        +textobjects
+channel         -hangul_input    +netbeans_intg   +timers
+cindent         +iconv           +num64           +title
+clientserver    +insert_expand   +packages        +toolbar
+clipboard       +job             +path_extra      +user_commands
+cmdline_compl   +jumplist        +perl/dyn        +vertsplit
+cmdline_hist    +keymap          +persistent_undo +virtualedit
+cmdline_info    +lambda          +postscript      +visual
+comments        +langmap         +printer         +visualextra
+conceal         +libcall         +profile         +viminfo
+cryptv          +linebreak       +python/dyn      +vreplace
+cscope          +lispindent      +python3/dyn     +wildignore
+cursorbind      +listcmds        +quickfix        +wildmenu
+cursorshape     +localmap        +reltime         +windows
+dialog_con_gui  +lua/dyn         +rightleft       +writebackup
+diff            +menu            +ruby/dyn        +X11
+digraphs        +mksession       +scrollbind      -xfontset
+dnd             +modify_fname    +signs           +xim
-ebcdic          +mouse           +smartindent     +xpm
+emacs_tags      +mouseshape      +startuptime     +xsmp_interact
+eval            +mouse_dec       +statusline      +xterm_clipboard
+ex_extra        +mouse_gpm       -sun_workshop    -xterm_save
+extra_search    -mouse_jsbterm   +syntax
+farsi           +mouse_netterm   +tag_binary
   system vimrc file: "/etc/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  system gvimrc file: "/etc/gvimrc"
    user gvimrc file: "$HOME/.gvimrc"
2nd user gvimrc file: "~/.vim/gvimrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
    system menu file: "$VIMRUNTIME/menu.vim"
  fall-back for $VIM: "/etc"
 f-b for $VIMRUNTIME: "/usr/share/vim/vim80"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_GTK  -pthread
-I/usr/include/gtk-3.0 -I/usr/include/atk-1.0
-I/usr/include/at-spi2-atk/2.0 -I/usr/include/pango-1.0
-I/usr/include/gio-unix-2.0/ -I/usr/include/cairo
-I/usr/include/gdk-pixbuf-2.0 -I/usr/include/glib-2.0
-I/usr/lib64/glib-2.0/include -I/usr/include/harfbuzz
-I/usr/include/freetype2 -I/usr/include/pixman-1 -I/usr/include/libpng15
-I/usr/include/libdrm     -O2 -g -pipe -Wall -fexceptions
-fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches
-m64 -mtune=generic -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64
-U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1
Linking: gcc   -L. -Wl,-z,relro -fstack-protector -rdynamic
-Wl,-export-dynamic -Wl,--enable-new-dtags
-Wl,-rpath,/usr/lib64/perl5/CORE  -Wl,-z,relro  -L/usr/local/lib
-Wl,--as-needed -o vim   -lgtk-3 -lgdk-3 -latk-1.0 -lgio-2.0
-lpangocairo-1.0 -lgdk_pixbuf-2.0 -lcairo-gobject -lpango-1.0 -lcairo
-lgobject-2.0 -lglib-2.0   -lSM -lICE -lXpm -lXt -lX11 -lSM -lICE  -lm
-lnsl  -lselinux   -lncurses -lacl -lattr -lgpm -ldl
-Wl,--enable-new-dtags -Wl,-rpath,/usr/lib64/perl5/CORE
-fstack-protector  -L/usr/lib64/perl5/CORE -lperl -lresolv -lnsl -ldl
-lm -lcrypt -lutil -lpthread -lc
```

**Comment by mgedmin** [comment]
For the record, I cannot reproduce the warnings with GTK+ 3.20 on Ubuntu GNOME, vim 8.0.21.

**Comment by mcepl** [comment]
Yes, it depends on the compilation flags for the Gtk. Traditionally, Fedora/RHEL builds with a lot of debugging options switched on.

**Comment by chrisbra** [comment]
so it is not a problem of Vim. Closing then.

**Comment by mcepl** [comment]
WHAT??? It is quite definitely problem of vim, but it is hidden on Ubuntu! "I don't see it, there is no problem"?

**Comment by vim-ml** [comment]
I get gtk warnings all the time, from both gtk2 and gtk3, and I can
reproduce them reliably.  Just set confirm, so that when you try to quit a
modified buffer in gvim it'll throw up that dialog box asking if you want
to save the buffer -- that dialog box generates no less than 6 gtk warnings
-- very ugly but apparently harmless

**Comment by mcepl** [comment]
Apparently, it _might_ be the problem of too old Gtk. @vim-ml what's the version of your Gtk3?

**Comment by sjlongland** [comment]
For what it's worth, I've seen similar warnings to this with vim-8.0 (various versions) linked against both gtk-2.0 and 3.0.

Having seen the earlier comments in this issue, I went looking on the GTK+ mailing list for a means of silencing these warnings, since it was clear to me there's a reluctance to fix the underlying issue.  This proved fruitless, and so far I'm experimenting with vim-qt as an alternative to gvim, as at present, gvim is almost unusable due to its incorrect usage of GTK+ triggering warnings on the console.

In amongst the discussions, was this comment:

> The "mapped child but invisible parent" warning is a violation of the
> documented invariants for GTK+ widgets; a child widget being mapped
> requires that all its ancestors, up to the top level, are visible.
>
> If you're getting that warning it means that the application developer
> is doing something *very* wrong; considering how gvim abuses GTK+ as
> if it were an X11 toolkit from the '90s, that does not surprise me one
> bit.

These warnings are visible when gvim is linked against either gtk+-2.24.28 or gtk+-3.16.7.

Looking at the above linked thread, perhaps if the devs are unwilling to fix this, they should explicitly require gtk+3.18.  The messages are most definitely not harmless as they pollute stderr.

**Comment by nuko8** [comment]
> For what it's worth, I've seen similar warnings to this with vim-8.0 (various versions) linked against both gtk-2.0 and 3.0.
> 
> Having seen the earlier comments in this issue, I went looking on the GTK+ mailing list for a means of silencing these warnings, since it was clear to me there's a reluctance to fix the underlying issue. This proved fruitless, 

No wonder at all.   Probably, almost all devs who have some experience in writing non-trivial gtk apps actually know how to silence those warnings once for all.  But they usually don't do that because they don't like to miss any of possible *real* issues by thoughtlessly silencing warnings even if they know some of warnings are definitely spurious.

In short, they don't care much about silencing warnings.

> and so far I'm experimenting with vim-qt as an alternative to gvim, as at present, gvim is almost unusable due to its incorrect usage of GTK+ triggering warnings on the console.

You can always open an issue here when you find an issue with gvim.   What functionality is unusable?  Where is the incorrect usage of GTK+?

> 
> In amongst the discussions, was this comment:
> 
> The "mapped child but invisible parent" warning is a violation of the
> documented invariants for GTK+ widgets; a child widget being mapped
> requires that all its ancestors, up to the top level, are visible.

There're five conditions for a Window to be visible on screen, and the comment you quoted is one of them.  But where is the invisible mapped child?  If one cannot find any invisible mapped child there, it is quite natural for him to conclude that warnings indicating that are false, or to wait for other additional, useful information on that.

> 
> If you're getting that warning it means that the application developer
> is doing something very wrong; considering how gvim abuses GTK+ as
> if it were an X11 toolkit from the '90s, that does not surprise me one
> bit.

I wish this comment was more specific.  What is done very wrongly?  What abuse and where?  I hope someday a person who gave that insightful comment will send a nice patch to us. 

> These warnings are visible when gvim is linked against either gtk+-2.24.28 or gtk+-3.16.7.
> 
> Looking at the above linked thread, perhaps if the devs are unwilling to fix this, they should explicitly require gtk+3.18. The messages are most definitely not harmless as they pollute stderr.

Personally, I like to leave that to users' discretion.  They can silence warnings by modifying the source code for themselves if necessary.  Though Vim is charityware, people are permitted to do that freely.  For details, please check the license.

**Comment by sjlongland** [comment]
> You can always open an issue here when you find an issue with gvim. What functionality is unusable? Where is the incorrect usage of GTK+?

That, I'd have to read up on where GTK+ is being used in `gvim` and how it is incorrect for me to be able to tell you.  I've done GUIs using Qt and wxWidgets, not GTK+ directly, and am not familiar with the innards of software that does.  I passed on the comment verbatim, thinking maybe in addition to the warning message (which mentions a `GtkEventBox` inside of `GtkNotebook`), it provided a hint.  I don't know any more.

The bit that's unusable is launching `gvim` from the terminal and still being able to use the terminal.  `gvim` linked against GTK+-2.0 will spray a few messages out on `stderr` on start-up.  `gvim` linked against GTK+-3.0 sprays a lot more.  This is something that never was a problem in older versions of `gvim` and so represents a regression.

It pushes what I'm trying to do in that terminal session off the scrollback buffer and clutters up the messages I want or need to see.

As I say, I have found a solution with `vim-qt`.  That starts up without spraying anything on the terminal session where I launched it.  The bonus being that it's in a graphics toolkit I have familiarity with, so should be able to fix future issues there.

As far as leaving dependencies in the user's discretion… do we do that with other dependencies?  I'm thinking if `gvim` depends on the behaviour of some particular version of GTK+, it should explicitly state that and make this fact aware to package maintainers.  They are not mind readers.

**Comment by nuko8** [comment]
> That, I'd have to read up on where GTK+ is being used in gvim and how it is incorrect for me to be able to tell you. I've done GUIs using Qt and wxWidgets, not GTK+ directly, and am not familiar with the innards of software that does. I passed on the comment verbatim, thinking maybe in addition to the warning message (which mentions a GtkEventBox inside of GtkNotebook), it provided a hint. I don't know any more.

So you actually don't know what is gvim's "incorrect usage of GTK+."  But that sounded like your own statement, not that of someone else...

> The bit that's unusable is launching gvim from the terminal and still being able to use the terminal. gvim linked against GTK+-2.0 will spray a few messages out on stderr on start-up. gvim linked against GTK+-3.0 sprays a lot more. This is something that never was a problem in older versions of gvim and so represents a regression.

If you do call it a regression, specify the version or patch number which you think causes the regression and *clarify what feature stops functioning by that.*   That's what "regression" means.

Note that, the GTK+ 2 gvim code had hardly been changed for more than a decade and the GTK+ 3 gvim was introduced about a year ago.  I'm wondering how it can be possible for Vim patches to introduce a regression.  As such, if any, it should not be hard for people to specify the number and open an issue accordingly.

Anyway, as written somewhere here or in other threads already, I think that's not a gvim issue at any rate, but due to implementation updates in the toolkits, and app behaviors resulting from them are not at app programmers' disposal through the API because of the nature of the updates.  IMHO, that inconvenience is a necessary cost of relying on an external library which keeps changing. 

> As far as leaving dependencies in the user's discretion… do we do that with other dependencies? I'm thinking if gvim depends on the behaviour of some particular version of GTK+, it should explicitly state that and make this fact aware to package maintainers. They are not mind readers.

Package maintainers don't need to be mind readers at all.  Build and run vim.  Read warning messages, if any.  Take a note on that and notify their users of that.  While it's almost impossible for me to know exactly which GTK+ version a user chooses to build gvim, package maintainers know the exact version which they provide with their built gvim.   They should know better than us.

**Comment by sjlongland** [comment]
https://mail.gnome.org/archives/gtk-list/2017-April/msg00005.html ← there is the statement quoted, from a GTK+ dev.  If it were my comment, I would have said so up front.

The problem has been an issue ever since vim-8.0 was released, and in the betas too.  Do a Internet search on the error message, and you see that lots of people are getting it, with all sorts of creative ways of dealing with it.

Somewhere there's even a post of mine in the comments of The Register (prior to me deleting the account; it now shows as "Anonymous Coward") making reference to this.  I didn't bother reporting it here, because being a beta, I expected there'd be bumps in the road, you were probably aware of the warnings and concluded that GTK+3 support wasn't ready yet.

Now that vim-8 is stable, I'd expect things had been cleaned up.  Something got missed, and that's fine, let's find it and fix it. :-)

The problem that is faced is that `gvim` (stable version) causes crap (from my perspective) to be emitted to `stderr`.  I've said that several times now.  It never used to, which is why it is a regression.

`gvim` forks into the background, but then does not relinquish control over `stdin`/`stdout`/`stderr` after forking like a sensible forked process should.  Not a problem in itself, but then it does things that make GTK+ spray its crap on `stderr`.

If you were watching a process in that terminal session, or trying to edit a command in a shell, too bad, you just have to guess where the cursor should be or try and refresh the screen somehow.  The latter is usually not too difficult with full-screen `ncurses` apps, but it is a problem if it's something like `bash` or even a `telnet` session to an embedded device.

At the moment, I am caught in the middle of two warring parties.  The GTK devs that say it's a `gvim` problem, and the Vim devs that say it's a GTK problem.  To me, I see two problems:

* a graphics toolkit that wants to yell at the user for code that the user didn't write, in a manner that makes their lives difficult
* an application that, if they warnings are to be taken at face value, is bending the rules as to what the graphics toolkit allows

Is GTK+ being overly pedantic?  Perhaps.  Is `gvim` bending the rules?  Perhaps.  I have suggested to GTK+ devs that an environment-variable switch so that devs can get their warnings and users be left in peace.  By the same token, the warnings is there for good reason, and I get that.

I'd be nice if there was a little more information in that message so we could find *where* the offence is being committed… apparently `systemd` has some way of receiving this information but I am not running `systemd`.  We have to work with what we've got.

Now, as it happens, I've just opened up `gui_gtk_x11.c` and had a look:

```
RC=0 stuartl@vk4msl-mb /tmp/vim-8.0.0386 $ grep --line-number -R 'gtk_notebook_new\|gtk_event_box_new' .
./src/gui_gtk_x11.c:3690:	    event_box = gtk_event_box_new();
./src/gui_gtk_x11.c:4100:    gui.tabline = gtk_notebook_new();
./src/gui_gtk_x11.c:4129:	event_box = gtk_event_box_new();
```

So two places, we create `GtkEventBox` instances, but only one `GtkNotebook`.  Once is when creating the initial `GtkEventBox` during initialisation `gui_mch_init`:

```
    {
	GtkWidget *page, *label, *event_box;

	/* Add the first tab. */
# if GTK_CHECK_VERSION(3,2,0)
	page = gtk_box_new(GTK_ORIENTATION_VERTICAL, 0);
	gtk_box_set_homogeneous(GTK_BOX(page), FALSE);
# else
	page = gtk_vbox_new(FALSE, 0);
# endif
	gtk_widget_show(page);
	gtk_container_add(GTK_CONTAINER(gui.tabline), page);
	label = gtk_label_new("-Empty-");
	gtk_widget_show(label);
	event_box = gtk_event_box_new();
	gtk_widget_show(event_box);
# if GTK_CHECK_VERSION(3,0,0)
	g_object_set_data(G_OBJECT(event_box), "tab_num", GINT_TO_POINTER(1L));
# else
	gtk_object_set_user_data(GTK_OBJECT(event_box), (gpointer)1L);
# endif
# if !GTK_CHECK_VERSION(3,14,0)
	gtk_misc_set_padding(GTK_MISC(label), 2, 2);
# endif
	gtk_container_add(GTK_CONTAINER(event_box), label);
	gtk_notebook_set_tab_label(GTK_NOTEBOOK(gui.tabline), page, event_box);
    }
```

The other is when adding a new tabs in `gui_mch_update_tabline`:
```
	if (page == NULL)
	{
	    /* Add notebook page */
# if GTK_CHECK_VERSION(3,2,0)
	    page = gtk_box_new(GTK_ORIENTATION_VERTICAL, 0);
	    gtk_box_set_homogeneous(GTK_BOX(page), FALSE);
# else
	    page = gtk_vbox_new(FALSE, 0);
# endif
	    gtk_widget_show(page);
	    event_box = gtk_event_box_new();
	    gtk_widget_show(event_box);
	    label = gtk_label_new("-Empty-");
# if !GTK_CHECK_VERSION(3,14,0)
	    gtk_misc_set_padding(GTK_MISC(label), 2, 2);
# endif
	    gtk_container_add(GTK_CONTAINER(event_box), label);
	    gtk_widget_show(label);
	    gtk_notebook_insert_page(GTK_NOTEBOOK(gui.tabline),
		    page,
		    event_box,
		    nr++);
	}
```

I don't use the tab feature of `gvim`, so I'm betting the former is the culprit, although given the similarity in the code, the other instance probably does the same thing.

Now, not being familiar with GTK+ as I mentioned, there are some hypotheses I would have to test:

* Is `gtk_widget_show` asynchronous?  Maybe there's a race condition here between us showing the tab page and its underlying child.
* Where do these pages get hidden?  I don't see a corresponding call to `gtk_widget_hide` when a different page is selected?
* Is it a case of a widget being already "shown" triggering this warning?

There isn't much code to traverse thankfully, and while okay, maybe the code hasn't changed in years, maybe it is changes introduced in GTK+ land that have caused this (aka, bitrot), I still think we should try and work together and fix this.  It is a problem that needs addressing as it does impact on the editor's usability.

**Comment by sjlongland** [comment]
Did some further experimentation, mousing over the toolbar buttons or navigating menus triggers this warning.  Repeatedly.

If I then open a file in a new tab, I get the warning twice more, then it falls silent.

**Comment by nuko8** [comment]
Probably, all you need to do is to spot a mapped, invisible child the existence of which is indicated by the warnings.  Where is that child?  Since the number of the widgets is finite, you can get it by looking up one by one, if any.  It's just that simple; no need for a huge pile of circumstantial evidence.

Note that, according to the reports we've received so far,  the warning are only seen with 3.16 or earlier, and that the GTK+3 gvim was developed using 3.18.  According to other reports, the warnings can't be seen with GTK+ versions higher than that.  We've also received some reports that other different warnings which were once seen with older GTK+ versions went away when the library was updated to a newer version, meanwhile we didn't do anything for that on our GUI code.

Why do the warnings go away as of 3.18?  If 3.16 were correct, we would still have them with 3.18, 3.20, and 3.22.   A regression of 3.18?   No, I don't think so.

That's the reason why I think the issue is not worth spending time.

**Comment by tonymec** [comment]
IIUC, it wasn't a *regression* in GTK 3.18 but a *fix* somewhere between GTK 3.16 and GTK 3.18.

FWIW, Vim 8.0.22 is already a little long in the teeth (the latest patchlevel as of this writing is 8.0.535) and the current stable release (Leap 42.2) of openSUSE Linux was published with GTK3 3.20.9 and later updated to 3.20.10. What is the latest GTK3 version available on the current release of Ubuntu? If you can get a GTK3 3.20, then by just installing it, and recompiling Vim with it (a full compile, e.g. "make reconfig" with configuration settings set by means of environment variables, then "make install") your problem will probably be fixed.
